execution/executor/src/components/in_memory_state_calculator_v2.rs

```diff
@@ -0,0 +1,229 @@
+    account_config::CORE_CODE_ADDRESS,
+    account_view::AccountView,
+    epoch_state::EpochState,
+    state_store::{
+        state_key::StateKey, state_storage_usage::StateStorageUsage, state_value::StateValue,
+    },
+    transaction::Transaction,
+    write_set::TransactionWrite,
+};
+use arr_macro::arr;
+use dashmap::DashMap;
+use rayon::prelude::*;
+use std::collections::HashMap;
+
+type ShardedStates = [DashMap<StateKey, Option<StateValue>>; 16];
+
+struct CoreAccountStateView<'a> {
+    base: &'a ShardedStates,
+    updates: &'a HashMap<&'a StateKey, &'a Option<StateValue>>,
+}
+
+impl<'a> CoreAccountStateView<'a> {
+    pub fn new(
+        base: &'a ShardedStates,
+        updates: &'a HashMap<&'a StateKey, &'a Option<StateValue>>,
+    ) -> Self {
+        Self { base, updates }
+    }
+}
+
+impl<'a> AccountView for CoreAccountStateView<'a> {
+    fn get_state_value(&self, state_key: &StateKey) -> Result<Option<Vec<u8>>> {
+        if let Some(v_opt) = self.updates.get(state_key) {
+            return Ok(v_opt.as_ref().map(|x| x.bytes().to_vec()));
+        }
+        if let Some(v_opt) = self.base[state_key.get_shard_id() as usize]
+            .get(state_key)
+            .as_ref()
+        {
+            return Ok(v_opt.as_ref().map(|x| x.bytes().to_vec()));
+        }
+        Ok(None)
+    }
+
+    fn get_account_address(&self) -> Result<Option<AccountAddress>> {
+        Ok(Some(CORE_CODE_ADDRESS))
+    }
+}
+
+/// Helper class for calculating state changes after a block of transactions are executed.
+pub struct InMemoryStateCalculatorV2 {}
+
+impl InMemoryStateCalculatorV2 {
+    pub fn calculate_for_transaction_block(
+        base: &StateDelta,
+        state_cache: StateCache,
+        to_keep: &[(Transaction, ParsedTransactionOutput)],
+        new_epoch: bool,
+    ) -> Result<(
+        Vec<HashMap<StateKey, Option<StateValue>>>,
+        Vec<Option<HashValue>>,
+        StateDelta,
+        Option<EpochState>,
+    )> {
+        ensure!(!to_keep.is_empty(), "Empty block is not allowed.");
+        ensure!(
+            base.base_version == base.current_version,
+            "Base version {:?} is different from current_version {:?}, cannot calculate state.",
+            base.base_version,
+            base.current_version,
+        );
+        ensure!(
+            base.updates_since_base.is_empty(),
+            "Updates is not empty, cannot calculate state."
+        );
+
+        let num_txns = to_keep.len();
+        for i in 0..num_txns {
+            let (txn, txn_output) = &to_keep[i];
+            ensure!(
+                Self::need_checkpoint(txn, txn_output) ^ (i != num_txns - 1),
+                "Checkpoint is allowed iff it's the last txn in the block. index: {i}, is_last: {}, txn: {txn:?}, is_reconfig: {}",
+                i == num_txns - 1,
+                txn_output.is_reconfig()
+            );
+        }
+
+        let StateCache {
+            ///// These don't change during the calculation.
```

Comment:
this line is irrelevant (used to describe two fields below it, in the declaration)

Reply:
Removed.